domain/machine/state/machine_cloud_instance.go

```diff
@@ -192,13 +194,192 @@ WHERE m.name = $machineName.name;
+	uuidQueryStmt, err := st.Prepare(uuidQuery, machineNameParam, machineUUID)
+	if err != nil {
+		return status.StatusInfo{}, errors.Trace(err)
+	}
+
+	// Prepare query for machine cloud instance status
+	instanceStatusParam := machineInstanceStatus{}
+	statusQuery := `SELECT &machineInstanceStatus.* FROM machine_cloud_instance_status WHERE machine_uuid = $machineUUID.uuid`
+	statusQueryStmt, err := st.Prepare(statusQuery, machineUUID, instanceStatusParam)
+	if err != nil {
+		return status.StatusInfo{}, errors.Trace(err)
+	}
+
+	// Prepare query for machine cloud instance status data
+	statusDataQuery := `SELECT &machineInstanceStatusData.* FROM machine_cloud_instance_status_data WHERE machine_uuid = $machineUUID.uuid`
+	statusDataQueryStmt, err := st.Prepare(statusDataQuery, machineUUID, machineInstanceStatusData{})
+	if err != nil {
+		return status.StatusInfo{}, errors.Trace(err)
+	}
+
+	var instanceStatusData []machineInstanceStatusData
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Query for the machine uuid
+		err := tx.Query(ctx, uuidQueryStmt, machineNameParam).Get(&machineUUID)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return errors.NotFoundf("machine %q", mName)
+			}
+			return errors.Annotatef(err, "querying uuid for machine %q", mName)
+		}
+
+		// Query for the machine cloud instance status
+		err = tx.Query(ctx, statusQueryStmt, machineUUID).Get(&instanceStatusParam)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return errors.Annotatef(machineerrors.StatusNotSet, "machine: %q", mName)
+			}
+			return errors.Annotatef(err, "querying machine status for machine %q", mName)
+		}
+
+		// Query for the machine cloud instance status data, no need to return
+		// error if we don't have any status data.
+		err = tx.Query(ctx, statusDataQueryStmt, machineUUID).GetAll(&instanceStatusData)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Annotatef(err, "querying machine status data for machine %q", mName)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return status.StatusInfo{}, errors.Trace(err)
+	}
+
+	// Transform the status data slice into a status.Data map.
+	statusDataResult := transform.SliceToMap(instanceStatusData, func(d machineInstanceStatusData) (string, interface{}) {
+		return d.Key, d.Data
+	})
+
+	instanceStatus := status.StatusInfo{
+		Message: instanceStatusParam.Message,
+		Since:   instanceStatusParam.Updated,
+		Data:    statusDataResult,
+	}
+
+	// Convert the internal status id from the (instance_status_values table)
+	// into the core status.Status type.
+	switch instanceStatusParam.Status {
+	case 0:
+		instanceStatus.Status = status.Empty
+	case 1:
+		instanceStatus.Status = status.Allocating
+	case 2:
+		instanceStatus.Status = status.Running
+	case 3:
+		instanceStatus.Status = status.ProvisioningError
+	}
+	return instanceStatus, nil
```

Comment:
I'd advocate for moving all the type transformations of the returning values from the db to state/types.go.